File: src/ui/shared/model/pid.py
from __future__ import annotations


class PID:
    def __init__(self, kp: float = 0.0, ki: float = 0.0, kd: float = 0.0) -> None:
        self._kp = float(kp)
        self._ki = float(ki)
        self._kd = float(kd)
        self._prev_error = 0.0
        self._integral = 0.0

    def update(self, error: float) -> float:
        self._integral += error
        derivative = error - self._prev_error
        self._prev_error = error
        return self._kp * error + self._ki * self._integral + self._kd * derivative

    def reset(self) -> None:
        self._prev_error = 0.0
        self._integral = 0.0

File: src/ui/react_ui/frontend/src/components/pid.ts
/**
 * PID controller — mirrors src/ui/shared/model/pid.py
 */
export class PID {
  private readonly kp: number;
  private readonly ki: number;
  private readonly kd: number;
  /** @internal */ _prevError = 0;
  /** @internal */ _integral = 0;

  constructor(kp: number, ki: number, kd: number) {
    this.kp = kp;
    this.ki = ki;
    this.kd = kd;
  }

  update(error: number): number {
    this._integral += error;
    const derivative = error - this._prevError;
    this._prevError = error;
    return this.kp * error + this.ki * this._integral + this.kd * derivative;
  }

  reset(): void {
    this._prevError = 0;
    this._integral = 0;
  }
}
